### DISCUSSION TITLE
Some questions

### DISCUSSION THREAD

**Discussion by c-ankur** [americanexpress/unify-flowret#16] [General] [answered]
1. Does the framework automatically execute each step and continues to the next step and execute it, or does it go into a pend state after each step by default? --> I saw in junits that after starting a case, there is a while loop to resumeCase. What's the reason for this?
2. What's the best way to pass the dynamic user/application data to each step or to the case itself so that all the steps get that data? Are process variables the only way to add dynamic / application/ request data? ( user_data seems to be a static field for every step. Can we add/change it dynamically?)

**Comment by deepakarora3** [comment]
Hi,

I would encourage you to read the documentation. All the answers to your questions are provided there. Plus it would also help to have more understanding of workflow and orchestration systems. Please do not take it as otherwise but these are really fundamental questions you are asking and which are related to workflow.

Does the framework automatically execute each step and continues to the next step and execute it, or does it go into a pend state after each step by default?
Flowret executes each step as per the definition. It is up to the step say whether it should proceed ahead or pend (due to an error or whatever). Flowret does not take this decision on its own.

 I saw in junits that after starting a case, there is a while loop to resumeCase. What's the reason for this?
These are unit test cases in which we are free to test different scenarios. The resume case is to check different scenarios.

What's the best way to pass the dynamic user/application data to each step or to the case itself so that all the steps get that data?
You could use process variables. Or the application could maintain a separate database based on the case id.

Are process variables the only way to add dynamic / application/ request data? ( user_data seems to be a static field for every step. Can we add/change it dynamically?)
Process variables are the only way to pass data and for steps to be able to change the data. User data is only additional information defined at definition time and which cannot be changed dynamically.

Most of the questions you are asking are addressed in the readme documentation.
Repli

**Comment by c-ankur** [comment]
I am indeed aware of workflow and orchestration systems. The reason I asked the question is that Junits were a bit misleading, and this being another implementation, I wanted to confirm. (Yes, the documentation mentions the execution of steps and how the response is used to carry on to the next step. Infinite while loops in junits for resume distracted me from the fact)

What's the best way to change process variables dynamically? Based on my needs, I need to pass a custom object inside the process variables. Looking at the code, it seems that the process variables types can be **only** Integer, string, boolean and long. I am hoping that it works with custom objects as well.

**Reply by deepakarora3** [reply, marked as answer]
What's the best way to change process variables dynamically?
Yes this is one thing I may have missed out in the documentation - will update. The process variables can be updated by first getting the process variables from the process context and then using the setValue method.

Based on my needs, I need to pass a custom object inside the process variables. Looking at the code, it seems that the process variables types can be only Integer, string, boolean and long. I am hoping that it works with custom objects as well.
Well, in my opinion, the process variables are not expected to pass Java objects. They are more intended for storing certain fields / values that drive orchestration. They are similar to exposed columns / fields in other BPM tools. Think about it this way - there may be hundreds or thousands of fields for a case. You would not want to keep all those fields as process variables. Only a few fields on which you drive some orchestration rules though even that is not a necessity. Hence the decision to have only certain types of values that can be stored. Also, when you talk of a Java custom object, you would have to create that object somewhere. That would most likely be part of your application as there is no way in Flowret to create a custom object. And so if you are creating it in the application then why move it to Flowret as a process variable. Just store in in your application and use it when required. Hope this clarifies.
